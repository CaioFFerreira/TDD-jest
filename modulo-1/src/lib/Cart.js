import find from 'lodash/find';
import remove from 'lodash/remove';
import Dinero from 'dinero.js';

const Money = Dinero;

export default class Cart {
 items = []

  add(item) {
    const itemTofind = { product: item.product};
    if(find(this.items, itemTofind)) {
      remove(this.items, itemTofind);
    }
    this.items.push(item);
  }

  remove(product) {
    remove(this.items, { product })
  }


  getTotal() {
   return this.items.reduce( (acc, item) => {
     return acc + item.quantity * item.product.price;
    },0)
  }
  sumary() {
    const total = this.getTotal()
    const items = this.items

    return {
      total,
      items,
    }
  }

  checkout() {
    const { total, items } = this.sumary()

    this.items = []

    return {
      total,
      items,
    }
  }

}