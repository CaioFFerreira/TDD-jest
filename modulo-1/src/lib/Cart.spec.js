import Cart from "./Cart.js";

describe("Cart", () => {
  let cart;
  let product = {
    title: "Adidas",
    price: 35388,
  };

  let product2 = {
    title: "Adidas - woman",
    price: 41872,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe("getTotal", () => {
    it("Should return 0 whan getTotal() is executed in a newly created", () => {
      expect(cart.getTotal()).toEqual(0);
    });

    it("Should multiply quantity and price and receive the total  amount", () => {
      const item = {
        product,
        quantity: 2, // 70776
      };
      cart.add(item);
      expect(cart.getTotal()).toEqual(70776);
    });

    it("Should ensure no more than on product exists at a time", () => {
      cart.add({
        product,
        quantity: 2, // 70776
      });
      cart.add({
        product,
        quantity: 1, // 35388
      });
      expect(cart.getTotal()).toEqual(35388);
    });

    it("Should update total when a product gets included and then removed", () => {
      cart.add({
        product,
        quantity: 2, // 70776
      });

      cart.add({
        product: product2,
        quantity: 1, // 70776
      });

      cart.remove(product);

      expect(cart.getTotal()).toEqual(41872);
    });

    
  });

  describe("checkout", () => {
    it("should return an object with the total and the list of items", () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 2,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });
    it("should reset the cart whan checkout() is called", () => {
      cart.add({
        product,
        quantity: 3,
      });
      cart.checkout();

      expect(cart.getTotal()).toEqual(0);
    })

    it("should return an object with the total and the list of items when summay is called", () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 2,
      });

      expect(cart.sumary()).toMatchSnapshot();
      expect(cart.getTotal()).toBeGreaterThan(0);
    });
  });
});
