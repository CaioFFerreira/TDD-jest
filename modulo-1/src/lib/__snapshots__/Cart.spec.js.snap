// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout should reset the cart whan checkout() is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 3,
    },
  ],
  "total": 106164,
}
`;

exports[`Cart checkout should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas - woman",
      },
      "quantity": 2,
    },
  ],
  "total": 154520,
}
`;

exports[`Cart checkout should return an object with the total and the list of items when summay is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas - woman",
      },
      "quantity": 2,
    },
  ],
  "total": 154520,
}
`;

exports[`Cart checkout should return an object with the total and the list of items when summay is called 2`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas - woman",
      },
      "quantity": 2,
    },
  ],
  "total": 154520,
}
`;

exports[`Cart checkout should return an object with the total and the list of items when summay is callled 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas - woman",
      },
      "quantity": 2,
    },
  ],
  "total": 154520,
}
`;
